### PR TITLE
Feature/f 53 nutrition country view

### PR DESCRIPTION
**Pull request by ArmanpreetGhotra**
-> Implement Complete Nutrition View

### CODE REVIEW

File: src/app/Nutrition/NutritionLayer.tsx [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+}: NutritionChoroplethProps) {
+  const [regionData, setRegionData] = useState<GeoJSON.FeatureCollection | null>(null);
+  const [showStateChoropleth, setShowStateChoropleth] = useState<boolean>(false);
+  const [selectedCountryId, setSelectedCountryId] = useState<GeoJSON.FeatureCollection | undefined>(undefined);
+
+  const geoJsonRef = useRef<L.GeoJSON | null>(null);
+
+  const map = useMap();
+
+  const handleCountryClick = async (feature: Feature<Geometry, GeoJsonProperties>, bounds: L.LatLngBounds) => {
```
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionLayer.tsx:33]: this needs to go in NutritionLayerOperations

File: src/app/Nutrition/NutritionLayer.tsx
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionLayer.tsx:1]: please call the file according to component name

File: src/app/Nutrition/NutritionLayer.tsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+          map.fitBounds(bounds);
+        }
+      } catch (error) {
+        console.error('Error fetching country data:', error);
+      }
+    }
+    if (handleClick) handleClick(feature, bounds, map);
+  };
+
+  const features = (feature: Feature<Geometry, GeoJsonProperties>, layer: L.Layer) => {
```
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionLayer.tsx:55]: operations

File: src/app/Nutrition/NutritionLayer.tsx [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    });
+
+    if (onEachFeature) onEachFeature(feature, layer);
+
+    const content = tooltip ? tooltip.render(feature) : feature.properties?.adm0_name;
+    const div = document.createElement('div');
+    const root = createRoot(div);
+    root.render(content);
+
+    layer.bindTooltip(div, { sticky: true, className: tooltip ? tooltip.className : 'countryTooltip' });
```
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionLayer.tsx:69]: I think country tooltip is out of scope for this task

File: src/app/Nutrition/NutritionLayer.tsx [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    <>
+      <GeoJSON ref={geoJsonRef} style={style} data={data} onEachFeature={features} />
+      <div className="absolute bottom-5 right-0 z-50 pr-10">
+        <LegendContainer items={[legendItem]} />
+      </div>
+      {showStateChoropleth && regionData && (
+        <StateChoropleth
+          regionNutri={selectedCountryId}
+          regionData={regionData}
+          handleClick={(feature: GeoJsonProperties) => console.log('State clicked:', feature)}
```
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionLayer.tsx:99]: unneded log

File: src/app/Nutrition/NutritionLayer.tsx [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+          tooltip={{
```
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionLayer.tsx:100]: state tooltip also not needed here

File: src/app/Nutrition/NutritionStateLayer.tsx [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+import { Layer, LeafletMouseEvent, PathOptions } from 'leaflet';
+import React, { useEffect, useRef, useState } from 'react';
+import { GeoJSON } from 'react-leaflet';
+
+import CustomAccordion from '@/components/Accordions/Accordion';
+import LegendContainer from '@/components/Legend/LegendContainer';
+import { GradientLegendContainerItem } from '@/domain/props/GradientLegendContainerItem';
+import NutritionStateChoroplethProps from '@/domain/props/NutritionStateProps';
+
+interface NutrientOption {
```
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionStateLayer.tsx:12]: this should be in domain

File: src/app/Nutrition/NutritionStateLayer.tsx [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+  label: string;
+  key: string;
+}
+
+interface AccordionItem {
```
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionStateLayer.tsx:17]: domain

File: src/app/Nutrition/NutritionStateLayer.tsx [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+  title: string;
+  iconSrc: string;
+  description: string;
+  content: JSX.Element;
+}
+
+type LayerWithFeature = Layer & {
+  feature?: GeoJSON.Feature<Geometry, GeoJsonProperties>;
+};
+const formatNutrientValue = (value: number | null | undefined): string => {
```
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionStateLayer.tsx:27]: move to operations or utils

File: src/app/Nutrition/NutritionStateLayer.tsx [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+  if (value === null || value === undefined) return 'N/A';
+  return `${value.toFixed(1)}%`;
+};
+
+const nutritionFill = (value: number | null): string => {
```
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionStateLayer.tsx:32]: operations

File: src/app/Nutrition/NutritionStateLayer.tsx [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        const nutrientLabels: { [key: string]: string } = {
+          fe_ai: 'Iron',
+          fol_ai: 'Folate',
+          va_ai: 'Vitamin A',
+          vb12_ai: 'Vitamin B12',
+          zn_ai: 'Zinc',
+          mimi_simple: 'Mean Adequacy Ratio',
+        };
+        const nutrientLabel = nutrientLabels[selectedNutrient] || 'Unknown Nutrient';
+        const tooltipContent = `
```
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionStateLayer.tsx:65]: I think it would be nice to have consistent styling for this tooltip. It should have dark/light bg color (depending on theme), have rounded corners and slightly transparent bg. You can take a look at tooltip styles in my pr

File: src/app/Nutrition/NutritionStateLayer.tsx [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+          <b>${feature.properties?.Name}</b><br/>
+          Risk of inadequate Intake of ${nutrientLabel}: ${formattedNutrientValue}
+        `;
+        layer.unbindTooltip();
+        layer.bindTooltip(tooltipContent, { className: tooltip?.className || 'state-tooltip' });
+      }
+    });
+  }, [selectedNutrient]);
+
+  const dynamicStyle = (feature: GeoJSON.Feature<Geometry, GeoJsonProperties> | undefined): PathOptions => {
```
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionStateLayer.tsx:75]: operations

File: src/app/Nutrition/NutritionStateLayer.tsx [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+
+    return {
+      fillColor: nutritionFill(value),
+      color: '#000',
+      weight: 1,
+      fillOpacity: 0.6,
+    };
+  };
+
+  const onEachFeature = (feature: GeoJsonProperties, layer: L.Layer): void => {
```
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionStateLayer.tsx:90]: operations

File: src/app/Nutrition/NutritionStateLayer.tsx
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionStateLayer.tsx:1]: call the file according to component name

File: src/app/Nutrition/NutritionStateLayer.tsx [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+      )}
+      <div className="absolute bottom-5 right-0 z-50 pr-10">
+        <LegendContainer items={[legendItem]} />
+      </div>
+      <GeoJSON data={regionData} style={dynamicStyle} onEachFeature={onEachFeature} />
+    </>
+  );
+}
+
+StateChoropleth.getAccordionData = function (
```
Comment (bohdangarchu) [inline src/app/Nutrition/NutritionStateLayer.tsx:153]: operations

**Review by bohdangarchu — CHANGES_REQUESTED** [review]
Great job! In addition to comments I noticed two things:
- I don't think we need this square border around the country
- on the first click half of the country is not shown
image

**Comment by ArmanpreetGhotra** [comment]
> Great job! In addition to comments I noticed two things:
> 
> * I don't think we need this square border around the country
> * on the first click half of the country is not shown
>   image

-> Strange, i never came across such issue, although i test it out multiple times and @plaume8  did testing too.

File: src/components/Buttons/CustomButton.tsx [hunk 14]
```diff
@@ -9,12 +9,17 @@ import { CustomButtonProps } from '@/domain/props/CustomButtonProps';
...
   const { variant } = attributes;
 
   return (
     <Button
       {...attributes}
+      style={{
+        backgroundColor: color,
+        minWidth: '120px',
+        textAlign: 'center',
+      }}
```
Comment (marinovl7) [inline src/components/Buttons/CustomButton.tsx:22]: This should be gone. First you should not be using style, until explicitly neccessary, but tailwind classes + clsx function . Nevertheless please do not change the Custom Button component without asking/communicating your special use case because this would change everything in the app

File: src/components/Map/Map.tsx [hunk 15]
```diff
@@ -1,14 +1,62 @@
...
+  const getFillColor = (dataType: string): string => {
+    switch (dataType) {
+      case 'actual':
+        return '#FFB74D';
+      case 'predicted':
+        return '#E3F2FD';
+      default:
+        return '#52525b';
+    }
+  };
```
Comment (marinovl7) [inline src/components/Map/Map.tsx:30]: this should not be here, operations

File: src/components/Map/Map.tsx [hunk 16]
```diff
@@ -1,14 +1,62 @@
...
+
+  // Fetch and preprocess data
+  useEffect(() => {
+    GlobalRepo.getNutritionData()
+      .then((global) => {
```
Comment (marinovl7) [inline src/components/Map/Map.tsx:35]: we are not using promise.then, instead use await + the data should not be fetched there if that is the use case, but in the parent component of the map, which is a server component. However this data could be fetched conditionally, if the `selectedMapType = "Nutrition"`, but again in the parent component

Comment (marinovl7) [inline src/components/Map/Map.tsx:35]: What is global -> use precise naming

File: src/components/Map/Map.tsx [hunk 17]
```diff
@@ -1,14 +1,62 @@
...
+                weight: 1,
+                fillOpacity: 0.5,
+                fillColor: getFillColor(item.data_type),
+              };
+              return acc;
+            },
+            {} as { [key: number]: L.PathOptions }
+          );
+
+          setCountryStyles(styles);
```
Comment (marinovl7) [inline src/components/Map/Map.tsx:50]: Why are you doing this ?

File: src/components/Map/Map.tsx [hunk 18]
```diff
@@ -25,6 +73,27 @@ export default function Map({ countries, disputedAreas }: MapProps) {
...
+            const style = countryStyles[c.properties.adm0_id];
+
+            return style ? (
+              <NutritionChoropleth
+                key={c.properties.adm0_id}
+                data={{ type: 'FeatureCollection', features: [c as Feature<Geometry, GeoJsonProperties>] }}
+                style={style}
+                handleClick={(sourceTarget) => {
+                  console.log('Country clicked:', sourceTarget);
+                }}
```
Comment (marinovl7) [inline src/components/Map/Map.tsx:89]: no need of this

File: src/components/Map/VectorTileLayer.tsx [hunk 19]
```diff
@@ -23,12 +25,11 @@ export default function VectorTileLayer({ countries, disputedAreas }: MapProps)
     );
     MapOperations.setMapInteractionFunctionality(baseMap);
     MapOperations.synchronizeLeafletMapbox(baseMap, mapContainer, context);
-
     return () => {
       baseMap.remove();
       context.map.off('move');
     };
-  }, [context, theme]);
+  }, [context, theme, selectedMapType]);
```
Comment (marinovl7) [inline src/components/Map/VectorTileLayer.tsx:32]: nope, this shouldn't go in here, this would cause the entire map to rerender when the map type changes, this should not be the case? Why did you need it?

File: src/app/Nutrition/NutritionStateLayer.tsx [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+
+  const legendItem: GradientLegendContainerItem = {
+    title: 'Risk of Inadequate Micronutrient Intake',
+    startColor: 'nutritionGradientStart',
+    endColor: 'nutritionGradientEnd',
+    startLabel: '0%',
+    endLabel: '100%',
+    middleColor: 'nutritionGradientMiddle',
+    tooltipInfo: 'Shows the inadequate ratio of nutrient intake.',
+  };
```
[the remark this answers is not in the document]
Comment (plaume8) [reply src/app/Nutrition/NutritionStateLayer.tsx]: will be done in F-53

File: src/app/page.tsx [hunk 21]
```diff
@@ -12,18 +12,25 @@ export default async function Home() {
   const countryMapDataPromise = globalRepo.getMapDataForCountries();
   const disputedAreasPromise = globalRepo.getDisputedAreas();
   const ipcDataPromise = globalRepo.getIpcData();
-  const [countryMapData, disputedAreas, ipcData] = await Promise.all([
+  const nutritionDataPromise = globalRepo.getNutritionData();
+  const [countryMapData, disputedAreas, ipcData, nutritionData] = await Promise.all([
```
Comment (plaume8) [inline src/app/page.tsx:16]: @marinovl7 or @bohdangarchu could you take a look and confirm if the `app/page` is the right place for the `CountryNutrion` data fetching ? -> for me it made sense to implement it there cause all other map related basic data is loaded there as well

**Comment by plaume8** [comment]
@bohdangarchu  @marinovl7 - to ensure that we can merge this issue as soon as possible, I’ve created a [new issue](https://linear.app/world-food-programme-wfp/issue/F-116/nutrition-map-use-reusable-legend) where I’ll refactor the legend to make sure @ahmedfarouk2000 reusable legend is used. By then, the changes from [this pull request](https://github.com/jst-seminar-rostlab-tum/wfp-hunger-map/pull/49/files) should also be merged, so the general legend will know which country is selected, which is also needed for the Nutrition view.

File: src/app/page.tsx (hunk 21 above)
Comment (bohdangarchu) [reply src/app/page.tsx:16]: looks good to me

File: src/components/Map/NutritionChoropleth.tsx [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+  const [countryStyles, setCountryStyles] = useState<{ [key: number]: L.PathOptions }>({});
+
+  // given the `CountryNutrition` data -> parse the data to map country styling
+  useEffect(() => {
+    const parsedStyles = NutritionChoroplethOperations.getCountryStyles(nutritionData);
+    setCountryStyles(parsedStyles);
+  }, [nutritionData]);
+
+  if (selectedAlert) {
+    toggleAlert(selectedAlert);
```
Comment (bohdangarchu) [inline src/components/Map/NutritionChoropleth.tsx:35]: this should only happen on country click. Right now this disables alerts for the global view and what we want is when you click on a country and zoom in then alerts are disabled

File: src/components/Map/NutritionStateChoropleth.tsx [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+          direction: 'top',
+          offset: [0, -10],
+          permanent: false,
+          sticky: true,
+        });
+      }
+    });
+  }, [selectedNutrient]);
+
+  const dynamicStyle = (feature: GeoJSON.Feature<Geometry, GeoJsonProperties> | undefined): PathOptions => {
```
Comment (bohdangarchu) [inline src/components/Map/NutritionStateChoropleth.tsx:57]: NutritionStateChoroplethOperations

File: src/components/Map/Map.tsx [hunk 24]
```diff
@@ -62,6 +70,22 @@ export default function Map({ countries, disputedAreas, ipcData }: MapProps) {
               toggleAlert={toggleAlert}
             />
           ))}
+      {selectedMapType === GlobalInsight.NUTRITION &&
+        countries.features &&
+        countries.features
+          .filter((countryData) => countryData.properties.interactive)
```
Comment (bohdangarchu) [inline src/components/Map/Map.tsx:76]: right now every country is clickable and zoom with region split works for every country. you can filter out only clickable countries by checking if their id is in `nutritionData`
image

**Review by bohdangarchu — CHANGES_REQUESTED** [review]
please fix these issues, otherwise lgtm

File: src/app/page.tsx (hunk 21 above)
Comment (Tschonti) [reply src/app/page.tsx:16]: Actually, it seems to me that we can delay fetching the nutrition data until the user selects the nutrition map. It's my first time looking at this PR, so feel free to challenge me if you think I'm wrong :D

in my understanding, the nutrition data fetched here is passed to the Map component, which passes it to `VectorTileLayer` and `NutritionChoropleth`. `VectorTileLayer` just passes it to the methods of `MapOperations`, but those methods don't use it at all. So we don't need to pass it to this component at all.

`NutritionChoropleth` is only rendered when the user selects the nutrition map, so this component can fetch the data. If you use the `useNutritionQuery` hook in the component's body, it'll automatically be fetched the first time this component loads. But in this case, we have to do something with the loading state (tho this request should be really fast). But generally, we are trying to avoid loading data before it's absolutely needed, so I recommend using this approach.

The current implementation of the IPC data required the data to be there from the beginning, that's why it was fetched on the server. But this is also being reworked.

**Comment by Tschonti** [comment]
The legend stays the same even when we're looking at the region view, I assume we should be using a gradient legend there

**Comment by marinovl7** [comment]
Please fix the merge conflicts, and I will merge this PR as it has been here for a long time! Create another ticket to implemented the required changes as the legend, the data fetching and etc.

**Comment by ArmanpreetGhotra** [comment]
@marinovl7 main branch is merged
